control-plane/src/modules/runs/index.ts

```diff
@@ -109,85 +96,69 @@ export const createRun = async ({
-          system_prompt: systemPrompt,
-          test,
-          test_mocks: testMocks,
-          reasoning_traces: reasoningTraces,
-          interactive: interactive,
-          enable_summarization: enableSummarization,
-          on_status_change: onStatusChange,
-          result_schema: resultSchema,
-          attached_functions: attachedFunctions,
-          agent_id: agentId,
-          agent_version: agentVersion,
-          model_identifier: modelIdentifier,
-          auth_context: authContext,
-          context: context,
-          enable_result_grounding: enableResultGrounding,
-        },
-      ])
-      .returning({
-        id: runs.id,
-        name: runs.name,
-        clusterId: runs.cluster_id,
-        systemPrompt: runs.system_prompt,
-        status: runs.status,
-        debug: runs.debug,
-        test: runs.test,
-        attachedFunctions: runs.attached_functions,
-        modelIdentifier: runs.model_identifier,
-        authContext: runs.auth_context,
-        context: runs.context,
-        interactive: runs.interactive,
-        enableResultGrounding: runs.enable_result_grounding,
-      });
-
-    run = result[0];
-
-    if (!!run && tags) {
-      await tx.insert(runTags).values(
-        Object.entries(tags).map(([key, value]) => ({
-          cluster_id: clusterId,
-          workflow_id: run!.id,
-          key,
-          value,
-        }))
-      );
-    }
-  });
+  // Insert the run with a subquery for debug value
+  const [run] = await db
+    .insert(runs)
+    .values({
+      id: runId ?? ulid(),
+      cluster_id: clusterId,
+      status: "pending",
+      user_id: userId ?? "SYSTEM",
+      ...(name ? { name } : {}),
+      debug: sql<boolean>`(SELECT debug FROM ${clusters} WHERE id = ${clusterId})`,
+      system_prompt: systemPrompt,
+      test,
+      test_mocks: testMocks,
+      reasoning_traces: reasoningTraces,
+      interactive: interactive,
+      enable_summarization: enableSummarization,
+      on_status_change: onStatusChange,
+      result_schema: resultSchema,
+      attached_functions: attachedFunctions,
+      agent_id: agentId,
+      agent_version: agentVersion,
+      model_identifier: modelIdentifier,
+      auth_context: authContext,
+      context: context,
+      enable_result_grounding: enableResultGrounding,
+    })
+    .returning({
+      id: runs.id,
+      name: runs.name,
+      clusterId: runs.cluster_id,
+      systemPrompt: runs.system_prompt,
+      status: runs.status,
+      debug: runs.debug,
+      test: runs.test,
+      attachedFunctions: runs.attached_functions,
+      modelIdentifier: runs.model_identifier,
+      authContext: runs.auth_context,
+      context: runs.context,
+      interactive: runs.interactive,
+      enableResultGrounding: runs.enable_result_grounding,
+    });
 
   if (!run) {
     throw new Error("Failed to create run");
   }
 
+  // Insert tags if provided
+  if (tags) {
+    await db.insert(runTags).values(
```

Comment:
Should still be possible to do this within the transaction?

Reply:
What do you think about not doing a transaction here? 

The tags would only be used in a small minority, whereas applying the transaction would apply to all.

We can do a `if (tags)` transaction, but that's a bit too much code?

Reply:
Is there any disadvantage to doing this transactionally though?

Reply:
I guess there would be some performance difference but that would be pretty marginal?

Reply:
I think the run row is extremely contentious right now, especially in the middle of a run. (asserting run ready etc). So, it's more to do with deadlocks and lags right now.

I think it's best to stay away from database transactions as much as you could, and instead rely on the database schema for atomic updates.

In hindsight, I think there's an opportunity here to reintegrate run tags into the workflow table.